### PR TITLE
Fixed async result storage with last modified header

### PR DESCRIPTION
**Pull request by Bladrak**
Following up #573

### CODE REVIEW

**Comment by Bladrak** [comment]
@masom seems we forgot the last-modified option when putting the result_storage async (given it's optional, it's not in the default result_storage)

**Comment by masom** [comment]
I wonder if we should update the `result_storage` to use a value object ( like storages )  that would contain some metadata.

Right now to get the last-modified value we make two calls to the result_storage when most backends support telling that value when the result is fetched.

**Comment by Bladrak** [comment]
Yes, I was wondering that as well when working on async support for aws connector, where I put a local cache within the class to avoid doubling the calls (not committed yet) as a temporary solution. However, updating the result_storage to return a value object instead of a binary might lead to important BC Breaks, wouldn't it?

**Comment by masom** [comment]
We would be able to support it with an `isinstance` check like in https://github.com/thumbor/thumbor/commit/7992b4a66b48b3f76acb3428b0f53a62efde7d01

If the return value isn't a `ResultStorageResult` then a call can be made for `last_modified`. That method could be flagged as `@deprecated`.

**Comment by Bladrak** [comment]
At the moment Storages don't seem to handle ValueObject, maybe we should have a common class to handle both ``StorageResult`` & ``ResultStorageResult``? Which could be common to ``LoaderResult`` maybe?

**Comment by masom** [comment]
I would assume storages and result_storages could use the same `StorageResult` object since there is a high chance they both provide the same kind of metadata.

`LoaderResult` should probably be separate from that ( the class is minimal and re-use not very efficient right now ).

As a start I would move the result_storage to return a `StorageResult` and also handle the current behaviour. A few `isinstance` will work until we can deprecate the old usage.

`StorageResult` could be based on the `LoaderResult` class.

**Comment by Bladrak** [comment]
Ok, I'll try to work on that then :)

**Comment by masom** [comment]
@Bladrak it should probably be a separate PR. This one should probably be merged.

**Comment by Bladrak** [comment]
Fine with me

File: thumbor/handlers/__init__.py [hunk 1]
```diff
@@ -250,16 +250,13 @@ def _load_results(self, context):
...
         if self.context.config.SEND_IF_MODIFIED_LAST_MODIFIED_HEADERS:
             # Handle If-Modified-Since & Last-Modified header
             try:
-                result_last_modified = self.context.modules.result_storage.last_updated()
-            except NotImplementedError:
-                logger.warn('last_updated method is not supported by your result storage service, hence If-Modified-Since & '
-                            'Last-Updated headers support is disabled.')
+                result_last_modified = yield gen.maybe_future(self.context.modules.result_storage.last_updated())
 
-            if result_last_modified:
```
Comment (masom) [inline thumbor/handlers/__init__.py]: Removing this if might make 
`if result_last_modified <= date_modified_since` succeed/ fail even if it should not.

if `result_last_modified` is `None` then if clause is always true.

**Comment by Bladrak** [comment]
@masom fixed

**Comment by masom** [comment]
:+1: Tested locally and it works.

**Comment by guilhermef** [comment]
@Bladrak @masom you guys are awesome

**Comment by heynemann** [comment]
The issue was created today, discussed today, pull requested today and merged today... Jeez I hope I'll be able to keep up with you guys! Awesome job! Really impressed. Great to have you both onboard!

**Pull request merged**